### PR TITLE
[CPDLP-3293] NPQExit - Configure logit

### PR DESCRIPTION
**Pull request by mooktakim**
### Context

Ticket: https://dfedigital.atlassian.net.mcas.ms/browse/CPDLP-3293

### Changes proposed in this pull request

* Added `append_info_to_payload` to `Api::BaseController` to track metadata for logit.

### Guidance

* [Search for the review app log in the last 10 min](https://kibana-uk1.logit.io/s/b2876053-0173-44fc-983d-99567292ba31/app/data-explorer/discover#?_a=(discover:(columns:!(app.host,url.original,url.path),isDirty:!t,sort:!()),metadata:(indexPattern:'8ac115c0-aac1-11e8-88ea-0383c11b333a',view:discover))&_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-10m,to:now))&_q=(filters:!(),query:(language:kuery,query:'*npq-registration-review-1568-*')))

### CODE REVIEW

File: app/controllers/api/base_controller.rb [hunk 1]
```diff
@@ -9,6 +9,57 @@ class BaseController < ActionController::API
...
+      payload[:current_user_class] = current_lead_provider&.class&.name
+      payload[:current_user_id] = current_lead_provider&.id
+      payload[:current_user_name] = current_lead_provider&.name
+
+      payload[:request_headers] = request.env.slice(
+        "CONTENT_TYPE",
+        "GATEWAY_INTERFACE",
+        "HTTP_ACCEPT",
+        "HTTP_ACCEPT_ENCODING",
+        "HTTP_AUTHORIZATION",
```
Comment (leandroalemao) [inline app/controllers/api/base_controller.rb:23]: Maybe we better off not including this as I saw: `"HTTP_AUTHORIZATION":"Bearer ambition-token"` in my tests on Logit?

File: app/controllers/api/base_controller.rb [hunk 2]
```diff
@@ -9,6 +9,57 @@ class BaseController < ActionController::API
...
+        "rack.url_scheme",
+        "rack.version",
+      ).transform_values(&:to_s).to_json
+      payload[:request_body] = request.raw_post
+
+      payload[:response_headers] = response.headers.transform_values(&:to_s).to_json
+      if response.status != 200
+        payload[:response_body] = response.body
+      end
+    end
```
Comment (leandroalemao) [inline app/controllers/api/base_controller.rb:61]: Seems other projects adding `request_query_params` to the payload as well.. maybe it's a good idea:

https://github.com/DFE-Digital/apply-for-teacher-training/blob/99dc097ac0c47678865ff2aa1f14dd3e2701ad81/app/controllers/api_docs/api_docs_controller.rb#L8

File: app/controllers/api/base_controller.rb [hunk 3]
```diff
@@ -9,6 +9,57 @@ class BaseController < ActionController::API
     rescue_from ActionController::BadRequest, with: :bad_request_response
     rescue_from ArgumentError, with: :bad_request_response
 
+    def append_info_to_payload(payload)
+      super
+      payload[:current_user_class] = current_lead_provider&.class&.name
+      payload[:current_user_id] = current_lead_provider&.id
+      payload[:current_user_name] = current_lead_provider&.name
+
+      payload[:request_headers] = request.env.slice(
```
Comment (leandroalemao) [inline app/controllers/api/base_controller.rb:18]: Why do we need this list of headers? Wouldn't be better to add all headers regardless or just remove some like `HTTP_AUTHORIZATION`?

**Review by leandroalemao — COMMENTED** [review]
Hi @mooktakim .. I've tested and added some comments.

Noticed on Logit, for every single request I do to the review app, there're 2 entries in the logs. not sure why?

image

File: app/controllers/api/base_controller.rb (hunk 1 above)
Comment (mooktakim) [reply app/controllers/api/base_controller.rb:23]: I'll remove it

File: app/controllers/api/base_controller.rb (hunk 3 above)
Comment (mooktakim) [reply app/controllers/api/base_controller.rb:18]: I'll remove auth

**Comment by mooktakim** [comment]
> Hi @mooktakim .. I've tested and added some comments.
> 
> Noticed on Logit, for every single request I do to the review app, there're 2 entries in the logs. not sure why?
> 
> image

I think there's two entries because of one from nginx and another from rails.

File: app/controllers/api/base_controller.rb (hunk 2 above)
Comment (leandroalemao) [reply app/controllers/api/base_controller.rb:61]: Think the query_params are mixed into the `params_json` if a request_body is sent.. just wondering if make sense to have a specific field in the payload only for the `query_params`.

image

File: config/environments/review.rb [hunk 4]
```diff
@@ -1,7 +1,7 @@
 require Rails.root.join("config/environments/production")
 
 Rails.application.configure do
-  config.log_level = :debug
+  config.log_level = :info
```
Comment (leandroalemao) [inline config/environments/review.rb:4]: Just noticed we have 2 lines for `config.log_level` on `config/environments/production.rb`

**Review by leandroalemao — APPROVED** [review]

File: app/controllers/api/base_controller.rb [hunk 5]
```diff
@@ -10,6 +10,56 @@ class BaseController < ActionController::API
     rescue_from ArgumentError, with: :bad_request_response
     rescue_from API::Errors::FilterValidationError, with: :filter_validation_error_response
 
+    def append_info_to_payload(payload)
```
Comment (cwrw) [inline app/controllers/api/base_controller.rb:13]: is this something we can move into a concern, and could we add a test for this as well please?

File: app/controllers/api/base_controller.rb [hunk 6]
```diff
@@ -10,6 +10,56 @@ class BaseController < ActionController::API
...
+        "rack.attack.called",
+        "rack.attack.throttle_data",
+        "rack.hijack?",
+        "rack.multiprocess",
+        "rack.multithread",
+        "rack.request.query_hash",
+        "rack.request.query_string",
+        "rack.run_once",
+        "rack.url_scheme",
+        "rack.version",
```
Comment (cwrw) [inline app/controllers/api/base_controller.rb:53]: discussed in person let's select the following:
```
{"HTTP_HOST"=>"xxx",
   "HTTP_ACCEPT"=>"application/json",
   "HTTP_VERSION"=>"HTTP/1.1",
   "QUERY_STRING"=>"page%5Bpage%5D=35&page%5Bper_page%5D=100",
   "HTTP_USER_AGENT"=>"GuzzleHttp/6.5.5 curl/7.83.1 PHP/7.4.33"},
   ```

File: app/controllers/api/base_controller.rb (hunk 2 above)
Comment (mooktakim) [reply app/controllers/api/base_controller.rb:61]: Ok i've added `query_params`

File: config/environments/review.rb (hunk 4 above)
Comment (mooktakim) [reply config/environments/review.rb:4]: Good catch @leandroalemao !
Very strange.
I have changed it to what we have for ECF app.

File: app/controllers/api/base_controller.rb (hunk 5 above)
Comment (mooktakim) [reply app/controllers/api/base_controller.rb]: Moved to concern and added test

File: config/environments/production.rb [hunk 7]
```diff
@@ -74,15 +74,10 @@
...
 
-  # Logging
-  config.log_level = :info
-
   if ENV["RAILS_LOG_TO_STDOUT"].present?
-    config.rails_semantic_logger.add_file_appender = false
-
     $stdout.sync = true
-
-    config.semantic_logger.add_appender(io: $stdout, level: Rails.application.config.log_level, formatter: :json)
```
Comment (cwrw) [inline config/environments/production.rb:85]: can we keep this, can see it's not defaulting to `Rails.application.config.log_level` unless I'm wrong?

Comment (cwrw) [reply config/environments/production.rb:85]: I can see it defaults to `:trace`

File: app/controllers/api/base_controller.rb [hunk 8]
```diff
@@ -4,34 +4,13 @@ class BaseController < ActionController::API
 
     include API::TokenAuthenticatable
     include ActionController::MimeResponds
+    include API::LoggerPayload
```
Comment (cwrw) [inline app/controllers/api/base_controller.rb:7]: might need to use prepend

**Review by cwrw — APPROVED** [review]

**Review by cwrw — APPROVED** [review]

**Pull request merged**